### PR TITLE
Refactor data and network logic into a single layer

### PR DESCRIPTION
**Pull request by yongla**
## Summary of Changes
- Network client, database, and `SharedPreferences` are now under a single point of access called `DataRepository`
- Course schedules are fetched when required instead of storing all subject specific schedules (prev. it was just storing all ECE schedules)
- Cleaned up unnecessary code

### CODE REVIEW

File: app/src/main/java/com/example/peterchu/watplanner/Networking/ApiInterface.java [hunk 1]
```diff
@@ -15,28 +15,21 @@
...
-    @GET("terms/{termId}/{subject}/{catalogNumber}/schedule.json")
+    @GET("courses/{subject}/{catalogNumber}/schedule.json")
     Call<CourseScheduleResponse> getCourseSchedule(
-            @Path("termId") String termId,
             @Path("subject") String subject,
             @Path("catalogNumber") String catalogNumber,
             @Query("key") String apiKey
     );
 
-    @GET("terms/{termId}/{subject}/schedule.json")
```
Comment (timothytong) [inline app/src/main/java/com/example/peterchu/watplanner/Networking/ApiInterface.java:28]: What happened to this?

Comment (yongla) [reply app/src/main/java/com/example/peterchu/watplanner/Networking/ApiInterface.java:28]: Previously we were just retrieving all course schedules under a certain subject (e.g. ECE) and storing that into the DB. Instead, we just retrieve the single course schedule when we need it: Viewing course details.

Down the the line, I'm thinking if we save a course we also store it in the DB

Comment (timothytong) [reply app/src/main/java/com/example/peterchu/watplanner/Networking/ApiInterface.java:28]: So when are we retrieving the schedule precisely? When we open course detail? Is that the only scenario?

Comment (yongla) [reply app/src/main/java/com/example/peterchu/watplanner/Networking/ApiInterface.java:28]: Yeah, that's the only time it's needed to be retrieved. You can view the call in `CourseDetailPresenter`

**Comment by Mrez95** [comment]
DAMN. 💣 - ship this beaut asap.

**Comment by timothytong** [comment]
LGTM

**Comment by PeterChu2** [comment]
LGTM!

**Pull request merged**